### DISCUSSION TITLE
SetPC() not doing what i expect

### DISCUSSION THREAD

**Discussion by RR7** [dirkwhoffmann/Moira#25] [Q&A] [answered]
Hi, i'm trying to implement Moira in Atari ST emulation, i have a few issues, i'll tackle them one at a time. firstly, memory is initialized before reset and moira correctly reads the reset vector correctly as 0xfc0020, however on cold reset i need to change the PC to 0xfc0000, i can do this but although i use SetPC and SetPC0, and can confirm in the debugger that these are set, moira still executes at 0xfc0020 and this crashes. i can solve that specific case using the Read16OnReset, and thats fine, my concern is, why doesnt execution change when i change the PC? there are a few places i need to do this. i wondered if it was prefetch, but there seems to be no prefect clear commands i can run?

**Comment by dirkwhoffmann** [comment, marked as answer]
The issue is the prefetch queue, as you've already pointed out. To redirect program execution, Moira has a separate function inside the debugger (see MoiraDebugger.h):

```c++
    //
    // Changing state
    //

    // Continues program execution at the specified address
    void jump(u32 addr);
```

The function first sets the PC and updates the prefetch queue afterward:

```c++
void
Debugger::jump(u32 addr)
{
    moira.reg.pc = addr & ~1;
    moira.fullPrefetch<C68000, POLL>();
}
```

`jump` is a public function, so you can call it via `moira.debugger.jump(<addr>)`.

Your project sounds interesting. I'd like to see Moira inside an ST emulator someday.

**Reply by RR7** [reply]
that solves that problem, thanks! i'd also like to see moira in an ST emulator, and i'm going to do my best to make it work. right now i'm getting 10 bombs on startup, so its alive. i'm going to try and setup hooks for illegal instruction, trap and line a. with these in place i should be able to get it working.

**Reply by mithrendal** [reply]
I also like to see it… hope you can bring it on. Is there already a public repo to give a star on ?

**Reply by RR7** [reply]
no repo yet, still early days :)
